attach/nv_attach_impl/pass/ptxpass_core/src/core.cpp

```diff
@@ -0,0 +1,435 @@
+	oss << std::cin.rdbuf();
+	if (!std::cin.good() && !std::cin.eof()) {
+		throw std::runtime_error("Failed to read from stdin");
+	}
+	return oss.str();
+}
+
+bool isWhitespaceOnly(const std::string &s)
+{
+	for (char c : s) {
+		if (!(c == ' ' || c == '\n' || c == '\r' || c == '\t' ||
+		      c == '\v' || c == '\f')) {
+			return false;
+		}
+	}
+	return true;
+}
+
+std::string getEnv(const char *key)
+{
+	const char *v = std::getenv(key);
+	return v ? std::string(v) : std::string();
+}
+
+std::pair<RuntimeInput, bool> parseRuntimeInput(const std::string &stdinData)
+{
+	RuntimeInput ri;
+	try {
+		auto j = json::parse(stdinData);
+		if (j.contains("full_ptx"))
+			ri.full_ptx = j["full_ptx"].get<std::string>();
+		if (j.contains("to_patch_kernel"))
+			ri.to_patch_kernel =
+				j["to_patch_kernel"].get<std::string>();
+		ri.global_ebpf_map_info_symbol = jsonGetString(
+			j, "global_ebpf_map_info_symbol", "map_info");
+		ri.ebpf_communication_data_symbol = jsonGetString(
+			j, "ebpf_communication_data_symbol", "constData");
+		return { ri, true };
+	} catch (...) {
+		ri.full_ptx = stdinData;
+		return { ri, false };
+	}
+}
+
+void emitRuntimeOutput(const std::string &outputPtx)
+{
+	json j;
+	j["output_ptx"] = outputPtx;
+	std::cout << j.dump();
+}
+
+// Validation: basic checks as placeholders
+bool validateInput(const std::string &input, const json &validation)
+{
+	if (validation.is_null())
+		return true;
+	if (validation.contains("require_entry") &&
+	    validation["require_entry"].get<bool>()) {
+		if (!containsEntryFunction(input))
+			return false;
+	}
+	if (validation.contains("require_ret") &&
+	    validation["require_ret"].get<bool>()) {
+		if (!containsRetInstruction(input))
+			return false;
+	}
+	if (validation.contains("ptx_version_min")) {
+		if (!validatePtxVersion(
+			    input,
+			    validation["ptx_version_min"].get<std::string>()))
+			return false;
+	}
+	return true;
+}
+
+bool containsEntryFunction(const std::string &input)
+{
+	return input.find(".visible .entry") != std::string::npos;
+}
+
+bool containsRetInstruction(const std::string &input)
+{
+	return input.find("\n    ret;") != std::string::npos ||
+	       input.find("\n\tret;") != std::string::npos;
+}
+
+bool validatePtxVersion(const std::string &input, const std::string &minVersion)
+{
+	// Expect line like: .version 7.0
+	std::istringstream iss(input);
+	std::string line;
+	double want = 0.0;
+	try {
+		want = std::stod(minVersion);
+	} catch (...) {
+		return true;
+	}
+	while (std::getline(iss, line)) {
+		if (line.rfind(".version", 0) == 0) {
+			// parse number
+			std::string v = line.substr(8);
+			try {
+				double have = std::stod(v);
+				return have >= want;
+			} catch (...) {
+				return true;
+			}
+		}
+	}
+	return true;
+}
+
+bool hasMarker(const std::string &input, const std::string &marker)
+{
+	return input.find(marker) != std::string::npos;
+}
+
+std::pair<std::string, bool> instrumentEntry(const std::string &input,
```

Comment:
Such functions should be put in targets like `ptxpass_kprobe_entry` or `ptxpass_kretprobe`